### PR TITLE
WIP Feat/multiselect

### PR DESCRIPTION
**Pull request by utkukaratas**
hi folks, this is still WIP but i'd be glad if you check it out 

instead of a seperate MultiSelect component, the SelectField is modified:

- .isMulti prop is introduced.
- SelectFieldValue type is introduced which will need to be used wherever the .value prop is handled. (check out the story changes)

so WDYT? if you guys think this is complex we could go the separate component way. not married to the idea as usual. 

cc @aigoncharov @jankoritak @danad02

### CODE REVIEW

File: packages/ui/src/SelectField.tsx
Comment (aigoncharov) [inline packages/ui/src/SelectField.tsx:52]: @utkukaratas I am not sure about this typing. Potentially, it allows users to pass value: V and the onChange that expects V[] back. It was one of the reasons I suggested to have MultiSelect as a separate component

File: packages/ui/src/SelectField.tsx
Comment (aigoncharov) [inline packages/ui/src/SelectField.tsx:120]: @utkukaratas still react-select allows us to use classes. Should we stick to classes and SCSS for consistency?

Comment (aigoncharov) [reply packages/ui/src/SelectField.tsx:120]: I think @jankoritak did a great job before with Select and demonstrated how it could be easily done with simple class overrides, or component replacements if necessary. WDYT?

File: packages/ui/src/SelectField.tsx
Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx:52]: hmm well if a type-happy coder like you finds it confusing we better seperate it :)) incoming...

File: packages/ui/src/SelectField.tsx
Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx:120]: @aigoncharov css was my first attempt too, but it's impossible to style the focused state of multiValue tags as per our hover-focus designs. i'll show you what i mean with an ss...

File: packages/ui/src/SelectField.tsx
Comment (aigoncharov) [reply packages/ui/src/SelectField.tsx:52]: Let's ask the team.

File: packages/ui/src/SelectField.tsx
Comment (aigoncharov) [reply packages/ui/src/SelectField.tsx:120]: If we cannot style the existing component for dinner reason, couldn't we replace it with our own one?
https://react-select.com/props#components

File: packages/ui/src/SelectField.tsx
Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx:52]: (context: do we create a seperate MultiSelect component **VS** augment SelectField with more intricate types to handle multiple selections)

Comment (jankoritak) [reply packages/ui/src/SelectField.tsx:52]: The current type seems good enough for common use-cases, so I'd be fine with keeping that. I don't have a strong preference, though.

File: packages/ui/src/SelectField.tsx
Comment (danad02) [reply packages/ui/src/SelectField.tsx:120]: I'd like to vote for replacement it even if the styling is possible. I find it cleaner.

File: packages/ui/src/SelectField.tsx
Comment (danad02) [reply packages/ui/src/SelectField.tsx:52]: Making almost identical component again because of this seems like an overkill to me. I know we want to be clean but maybe that's the case we're fine with. A comment might help if the user struggles with types.

Comment (danad02) [reply packages/ui/src/SelectField.tsx:52]: @utkukaratas @aigoncharov Maybe this is a solution?
```
export type SelectFieldCoreDynamicProps<V> =
  | {
      isClearable: true
      value: SelectFieldOption<V> | null
      onChange: (newValue: SelectFieldOption<V> | null) => void
    }
  | {
      isClearable?: false
      value: SelectFieldOption<V>
      onChange: (newValue: SelectFieldOption<V>) => void
    }
  | {
      isClearable: true
      value: SelectFieldOption<V>[] | null
      onChange: (newValue: SelectFieldOption<V>[] | null) => void
    }
  | {
      isClearable?: false
      value: SelectFieldOption<V>[]
      onChange: (newValue: SelectFieldOption<V>[]) => void
    }
```

Comment (aigoncharov) [reply packages/ui/src/SelectField.tsx:52]: @danad02 good idea!

Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx:52]: that's a mouthful but why not. thx @danad02. on its way...

**Comment by utkukaratas** [comment]
hi there @jankoritak,

we have an issue with **SelectField / "menuIsOpen=true" stories** when **onChange/setValue combo** is used on the Storyboard (see my latest commit pls). if you try to select something, you'll see the popup jump around. do you have any tips why this could happen. i better solve this while i'm on it.

**Comment by utkukaratas** [comment]
verify-all

**Comment by utkukaratas** [comment]
verify

File: packages/ui/src/SelectField.tsx
Comment (danad02) [inline packages/ui/src/SelectField.tsx:49]: Since `isMulti` is present in `SelectFieldCoreDynamicProps` we can remove it here.

File: packages/ui/__tests__/SelectField.test.tsx
Comment (danad02) [inline packages/ui/__tests__/SelectField.test.tsx:222]: Do we need this?

Comment (danad02) [inline packages/ui/__tests__/SelectField.test.tsx:222]: Do we need this?

Comment (utkukaratas) [reply packages/ui/__tests__/SelectField.test.tsx]: it seems it's gone from the master too, removing...

File: packages/ui/src/SelectField.tsx
Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx]: nice catch, thx!

**Comment by utkukaratas** [comment]
btw @jankoritak, i'll wait for your approval for the issue above before merging.

File: packages/ui/src/SelectField.tsx
Comment (jankoritak) [reply packages/ui/src/SelectField.tsx]: It seems like this one was not addressed yet. Can you look into it, @utkukaratas? Once this is addressed, it seems good for approval.

Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx]: @jankoritak this was an outdated convo. resolving it...

Comment (jankoritak) [reply packages/ui/src/SelectField.tsx]: Got it. Could you rebase the branch? AFAIC, the PR is good to go.

**Review by jankoritak — APPROVED** [review]

File: packages/ui/src/SelectField.tsx
Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx]: done @jankoritak

**Review by danad02 — APPROVED** [review]
LGTM. Nice job :+1:

File: packages/ui/src/SelectField.module.scss
Comment (aigoncharov) [inline packages/ui/src/SelectField.module.scss:103]:
```suggestion
    border: c.$borderWidth solid c.$colorAccessibilityOutline;
```

File: packages/ui/src/SelectField.module.scss
Comment (aigoncharov) [inline packages/ui/src/SelectField.module.scss:108]:
```suggestion
  border: c.$borderWidth solid c.$colorPrimary;
```

Comment (aigoncharov) [reply packages/ui/src/SelectField.module.scss:108]: Should we also add a `1px` transparent border for the unhovered state, so the cell does not jump when hovered?

File: packages/ui/__stories__/SelectField.stories.tsx
Comment (aigoncharov) [inline packages/ui/__stories__/SelectField.stories.tsx:322]: Should we wrap it with a `div` with some max-wdith, say, 400px, to make it really multiple rows? On the current screenshot it is only one row

File: packages/ui/src/SelectField.tsx
Comment (aigoncharov) [inline packages/ui/src/SelectField.tsx:37]: @utkukaratas besides forcused state of the chip, should we style its default state? I would suggest the following:
- Use one of our neutral colors for its backgorund
- Use typography helpers (typographyBodySmall?) to use standard font sizes
- Use colorText for the text color
- Replace remove icon with the one from react-feather, so it matches what we use in our `Toast` component.
- Optionnally, try matching padding and margins with what we have in our `Toast`component

WDYT?

File: packages/ui/src/SelectField.tsx
Comment (aigoncharov) [inline packages/ui/src/SelectField.tsx:64]: Should we also use the same technique in `SelectFieldFormikProps`? Could we also have a story for the formik integration of MultiSelect?

**Review by aigoncharov — CHANGES_REQUESTED** [review]
@utkukaratas left a couple of comments! Great job so far!

File: packages/ui/src/SelectField.tsx
Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx:41]: we now use our IconButton for the button. resolving...

File: packages/ui/src/SelectField.tsx
Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx:78]: this is done and committed. i'll request a review specifically for this.

**Review by danad02 — APPROVED** [review]

**Comment by utkukaratas** [comment]
hi @aigoncharov this is ready for a review. all of the issues above are handled.

**Comment by aigoncharov** [comment]
@utkukaratas I have tried the storybook. I cannot remove selected items by clicking their `x`. The click opens the dropdown menu instead.l

File: packages/ui/src/SelectField.tsx
Comment (aigoncharov) [reply packages/ui/src/SelectField.tsx:41]: @utkukaratas should we resolve all other proposals? I see that only the remove icon was changed.
I still suggest we do:
- Use one of our neutral colors for its backgorund
- Use typography helpers (typographyBodySmall?) to use standard font sizes
- Use colorText for the text color
- Try matching padding and margins with what we have in our Toastcomponent

WDYT?

**Comment by aigoncharov** [comment]
@utkukaratas could you also take a look at https://github.com/hazelcast/frontend-shared/pull/151#discussion_r568357866 ?

**Review by aigoncharov — CHANGES_REQUESTED** [review]
@utkukaratas I think we are getting really close! Left a couple of comments

File: packages/ui/src/SelectField.tsx
Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx:41]: @aigoncharov i've used our IconButton instead of doing those manually. much cleaner IMO. it seems recent IconButton changes broke this PR but now that's fixed too.

File: packages/ui/src/SelectField.tsx
Comment (aigoncharov) [inline packages/ui/src/SelectField.tsx:51]: Can't we stop propagation for the onClick and not introduce the mouseDown handler?

File: packages/ui/src/SelectField.tsx
Comment (aigoncharov) [reply packages/ui/src/SelectField.tsx:41]: @utkukaratas I see that you used IconButton before. I am talking abou the color, padding and typography of the chip that represents a selected item. Currently, the chip does not match our styling.

File: packages/ui/src/SelectField.tsx
Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx:51]: nope, that doesn't work. it was the MouseDown chain of events that should be stopped.

File: packages/ui/src/SelectField.tsx
Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx:41]: ah! i misunderstood that. will handle it...

Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx]: @aigoncharov 

- all of the issues above should be resolved. 
- it should be pixel perfect with the new design. (except some parts in the design those are not multiples of 4px)

**Comment by utkukaratas** [comment]
verify

**Comment by utkukaratas** [comment]
pr-builder seems to be stuck :/

**Comment by utkukaratas** [comment]
verify

**Comment by aigoncharov** [comment]
@utkukaratas it looks amazing! Great job! On last bit
image
Should we match the paddings for the placeholder text for an empty input?

File: packages/ui/src/SelectField.module.scss
Comment (aigoncharov) [inline packages/ui/src/SelectField.module.scss:114]: Good thinking! I totally agree with making it a sane increment of `grid` instead of being 100% pixel perfect

File: packages/ui/src/SelectField.module.scss
Comment (aigoncharov) [inline packages/ui/src/SelectField.module.scss:136]: nit
```suggestion
  .multiValueLabel {
    @include c.typographyBodySmall;
    @inclide h.overflowEllipsis;

    color: c.$colorText;
    padding: 0 c.$grid;
    box-sizing: border-box;
  }
```

File: packages/ui/src/SelectField.tsx
Comment (aigoncharov) [inline packages/ui/src/SelectField.tsx:39]: nit: SHould `MultiValue` come after `MultiValueRemove`and `MultiValueLabel` since it references them? AFAIU, `consts` are not hoisted all the way to the top, so it works only because these arrow functions are converted to regular functions during transpilation

**Comment by aigoncharov** [comment]
@utkukaratas clear all button is kind of misaligned. Could you take a look?
image

File: packages/ui/src/SelectField.tsx
Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx:39]: i don't get the point here??? MultiValue uses the other 2 components inside its body - so they will be referenced at the time of the call. declaration order should not be an issue here in any case.

File: packages/ui/src/SelectField.module.scss
Comment (utkukaratas) [reply packages/ui/src/SelectField.module.scss:114]: there're also some 1px mismatches between paddings and what not. (i'll demo this to @pawelklasa for his blessings to the implementation :D designer folks have ultra precise / lazer zoom eyes for these kind of things.)

File: packages/ui/src/SelectField.tsx
Comment (aigoncharov) [reply packages/ui/src/SelectField.tsx:39]: @utkukaratas here is a simple reproducer https://www.typescriptlang.org/play?noUnusedLocals=true&noUnusedParameters=true&declaration=false&target=99#code/MYewdgzgLgBAhjAvDAFASiQPhgbwFAwwBG6eAvnnKXqJLEUqhotrRCADYCmAdByAHMUAcgAWAW0nC0QA
Try running it and see the console

Comment (aigoncharov) [reply packages/ui/src/SelectField.tsx:39]: We get away with it because webpack transpiles it to https://www.typescriptlang.org/play?noUnusedLocals=true&noUnusedParameters=true&declaration=false&target=99#code/PTAEGMHsDsGcBdQENQF5QAoCUaB8oBvAKBFDICNsSwBfa5K+qORctTHVfZ2SAGwCmAOj6QA5hgDkACwC28yViL0AKgCckcAA4BLQQBNQ5AJ6gA7gPJak4ANah4kZQDMArtHDwdMZB0JEKKjokRjcPLx9KHGIyHn5hUQkZBSUaIA

Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx:39]: ok calling a() before declaring "b" might of course be problematic but in our case noone is calling MultiValue before declaring the other two. and MultiValue is instantiated much much later, in fact ages later, when SelectField is instantiated somewhere in the app code.

Comment (aigoncharov) [reply packages/ui/src/SelectField.tsx:39]: @utkukaratas yeah, you are right. Here is a great article that helped me to brush up my knowledge about lexical scoping BTW https://blog.bitsrc.io/hoisting-in-modern-javascript-let-const-and-var-b290405adfda

Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx:39]: @aigoncharov you're a review machine dude, and every machine needs a reboot at one point, you just got segfaulted :D

**Comment by utkukaratas** [comment]
setting the PR to draft. that placeholder margin issue opened such a can of worms :(((

**Comment by aigoncharov** [comment]
> setting the PR to draft. that placeholder margin issue opened such a can of worms :(((

Can;t we just add extra margin to the `.hz-select-field__placeholder` element when it is a child of a multi value select?

**Comment by utkukaratas** [comment]
> Can;t we just add extra margin to the `.hz-select-field__placeholder` element when it is a child of a multi value select?

yes but thet the input (blinking caret) container looks off. tried a few different ways. that resulted lots of cascading failures :/ i'll sleep on this to find a better way.

**Comment by aigoncharov** [comment]
@utkukaratas when container has a value it has `.hz-select-field__value-container--has-value` class applied. You can change the horizontal margin here https://github.com/hazelcast/frontend-shared/pull/151/files#diff-bc420743ebb0252dfa1d6d8e7cdfd29a04c885dbc6127186fe1a98d68537a83dR113 only when there is a value/values

**Comment by utkukaratas** [comment]
> @utkukaratas when container has a value it has `.hz-select-field__value-container--has-value` class applied. You can change the horizontal margin here https://github.com/hazelcast/frontend-shared/pull/151/files#diff-bc420743ebb0252dfa1d6d8e7cdfd29a04c885dbc6127186fe1a98d68537a83dR113 only when there is a value/values

thx, this has been a great trick! though that css is now a living organism on its own :/

i'll demo this to @pawelklasa tomorrow and then will submit.

**Comment by utkukaratas** [comment]
@aigoncharov got the approval from pawel; ready for code review.

File: packages/ui/src/SelectField.tsx
Comment (aigoncharov) [inline packages/ui/src/SelectField.tsx:71]: @utkukaratas why do we need this line?

Comment (aigoncharov) [reply packages/ui/src/SelectField.tsx:71]: Can't we just pass `onClick` to the button? BTW, should we just use `IconButton`?

Comment (aigoncharov) [reply packages/ui/src/SelectField.tsx:71]: Oh, I see now that it is custom styled.The question regarding `onClick` still stands.

Comment (aigoncharov) [reply packages/ui/src/SelectField.tsx:71]: @utkukaratas there is a small visual problem 
image

Comment (aigoncharov) [reply packages/ui/src/SelectField.tsx:71]: As you can see the hover background for the remove button overlaps with the focus shadow. Shouldn't it go underneath?

Comment (aigoncharov) [reply packages/ui/src/SelectField.tsx:71]: We'll probably need a transparent absolutely positioned element on top of the button to draw the shadow on the very top

Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx:71]: > @utkukaratas why do we need this line?

we need to pass the onclick given from the parent to our custom div. it's done in `mousedown`  to do everything in one fell swoop and keep the linter happy.

Comment (aigoncharov) [reply packages/ui/src/SelectField.tsx:71]: @utkukaratas personally, I would expect the removal to happen only when I release the mouse, but it is a nit. It is just that click is usually happens after mouse is up, so currently it feels a bit unnatural. WDYT? What error does a linter give you?

Comment (aigoncharov) [reply packages/ui/src/SelectField.tsx:71]: I tried it. It is a keybord listener error, right? I would suggest to just put a eslint ignore on this line.  WDYT?

Comment (utkukaratas) [reply packages/ui/src/SelectField.tsx:75]: > We'll probably need a transparent absolutely positioned element on top of the button to draw the shadow on the very top

just modified the box shadow to be NOT inset ;)

**Review by aigoncharov — APPROVED** [review]
Great job! let's merge this one in!

**Pull request merged**
